include/fiction/algorithms/simulation/sidb/quickexact.hpp

```diff
@@ -0,0 +1,506 @@
+                charge_lyt_copy.recompute_system_energy();
+
+                for (const auto& cell : detected_negative_sidbs)
+                {
+                    charge_lyt_copy.adding_sidb_to_layout(cell, -1);
+                }
+
+                result.charge_distributions.push_back(charge_lyt_copy);
+            }
+
+            charge_layout.reset_charge_index_sub_layout();
+
+            charge_layout.increase_charge_index_by_one(
+                dependent_cell_mode::VARIABLE, energy_calculation::KEEP_OLD_ENERGY_VALUE,
+                charge_distribution_history::CONSIDER,
+                exhaustive_algorithm::QUICKEXACT);  // "false" allows that the charge state of the dependent cell is
+                                                    // automatically changed based on the new charge distribution.
+        }
+
+        // Charge configurations of the sub-layout are iterated
+        while (charge_layout.get_charge_index_sub_layout().first < charge_layout.get_max_charge_index_sub_layout())
+        {
+            if (charge_layout.is_physically_valid())
+            {
+                charge_distribution_surface<Lyt> charge_lyt_copy{charge_layout};
+                charge_lyt_copy.recompute_system_energy();
+
+                // The previously detected negatively-charged SiDBs are added to the final layout.
+                for (const auto& cell : detected_negative_sidbs)
+                {
+                    charge_lyt_copy.adding_sidb_to_layout(cell, -1);
+                }
+
+                result.charge_distributions.push_back(charge_lyt_copy);
+            }
+
+            charge_layout.increase_charge_index_of_sub_layout_by_one(
+                dependent_cell_mode::VARIABLE, energy_calculation::KEEP_OLD_ENERGY_VALUE,
+                charge_distribution_history::CONSIDER, exhaustive_algorithm::QUICKEXACT);
+        }
+
+        if (charge_layout.is_physically_valid())
+        {
+            charge_distribution_surface<Lyt> charge_lyt_copy{charge_layout};
+
+            for (const auto& cell : detected_negative_sidbs)
+            {
+                charge_lyt_copy.adding_sidb_to_layout(cell, -1);
+            }
+
+            result.charge_distributions.push_back(charge_lyt_copy);
+        }
+
+        for (const auto& cell : detected_negative_sidbs)
+        {
+            layout.assign_cell_type(cell, Lyt::cell_type::NORMAL);
+        }
+    }
+    /**
+     * This function is required for the initialization.
+     */
+    void initialize_charge_layout()
+    {
+        if constexpr (has_get_sidb_defect_v<Lyt>)
+        {
+            layout.foreach_sidb_defect(
+                [this](const auto& cd)
+                {
+                    const auto& [cell, defect] = cd;
+
+                    if (defect.type != sidb_defect_type::NONE)
+                    {
+                        charge_lyt.add_defect(cell, layout.get_sidb_defect(cell));
```

Comment:
I'm not super happy with how defects are handled in the `charge_distribution_surface`. The `sidb_surface` class exists and holds defects. Why is there a need to essentially reimplement its functionality and copy all the defects over to the `charge_distribution_surface`?

Reply:
@marcelwa I agree that this seems weird and I am also not happy, but I am struggling to find a way that seems as efficient as this. I may be wrong, but this is why classes are huge.